exporter/launch_failures.py

```diff
@@ -0,0 +1,82 @@
+from base import OSBase
+from osclient import session_adapter
+from prometheus_client import CollectorRegistry, generate_latest, Gauge
+from utils import node_details
+from datetime import datetime, timedelta
+from collections import Counter
+import logging
+
+logger = logging.getLogger(__name__)
+
+LABELS = [
+    'region',
+    'project_id'
+    'project_name']
+
+
+class LaunchFailures(OSBase):
+    """Class to report the statistics on OpenStack Nodes"""
+
+    def __init__(self, oscache, osclient):
+        super(LaunchFailures, self).__init__(oscache, osclient)
+        self.refresh_interval = oscache.refresh_interval
+
+    def build_cache_data(self):
+        """Return list of stats to cache."""
+        nova_api = session_adapter('compute')
+        keystone_api = session_adapter('identity')
+
+        nodes = node_details.get_nodes(detail=True)
+        servers = nova_api.get(
+            'servers/detail?all_tenants=True').json()['servers']
+
+        projects = keystone_api.get('v3/projects').json()['projects']
+        project_names = {p['id']: p['name'] for p in projects}
+
+        instance_failures = [
+            (s['project_id'], project_names[s['project_id']])
+            for s in servers
+            if self._valid_failure(s['status'] == 'ERROR', s['updated'])]
+
+        node_failures = [
+            (n.project_id, n.project_name)
+            for n in nodes
+            if self._valid_failure(n.last_error is not None, n.updated)
+        ]
+
+        counts = Counter(instance_failures + node_failures)
```

Comment:
I haven't worked much with `Counter` (and in fact I was confused, thinking this was a Prometheus abstraction also called Counter.) So does it aggregate counts of unique items? Because we could be passing in multiple copies of a project id/name tuple here.

Reply:
Yep. It's a convenient abstraction for counting items in a list.